lib/tapioca/dsl/compilers/delegate.rb

```diff
@@ -0,0 +1,137 @@
+
+module Tapioca
+  module Dsl
+    module Compilers
+      # `Tapioca::Dsl::Compilers::Delegate` generates RBI files for classes that use the `delegate` method
+      # from ActiveSupport.
+      #
+      # For a class like:
+      #
+      # ```ruby
+      # class Delegator
+      #  sig { returns(Target) }
+      #  attr_reader :target
+      #
+      #  delegate :method, to: :target
+      # end
+      #
+      # class Target
+      #  sig { returns(String) }
+      #  def method = "hi"
+      # end
+      # ```
+      #
+      # This compiler will generate the following RBI file:
+      #
+      # ```rbi
+      # class Delegator
+      #  sig { returns(Target) }
+      #  attr_reader :target
+      #
+      #  sig { returns(String) }
+      #  def method; end
+      # end
+      # ```
+      #
+      # The `delegate` method can also take the `prefix`, `private` and `allow_nil` options but is not intelligent
+      # about discovering types from instance variables, class_variables and constants - if you delegate to a target
+      # whose type is not discoverable statically, the type will default to T.untyped
+      #
+      # Delegates that _themselves_ return a `T.untyped` value will not be generated in the RBI file, since Sorbet
+      # already generates a `T.untyped` return by default
+      #
+      class Delegate < Compiler
+        extend T::Sig
+
+        ConstantType = type_member { { fixed: T.all(::Module, Extensions::Module) } }
+
+        class << self
+          extend T::Sig
+
+          sig { override.returns(T::Enumerable[::Module]) }
+          def gather_constants
+            return [] unless defined?(Tapioca::Dsl::Compilers::Extensions::Module)
+
+            all_classes.select do |c|
+              c.singleton_class < Tapioca::Dsl::Compilers::Extensions::Module &&
+                T.unsafe(c).__tapioca_delegated_methods.any?
+            end
+          end
+        end
+
+        sig { override.void }
+        def decorate
+          root.create_path(constant) do |klass|
+            constant.__tapioca_delegated_methods.each do |delegated_method|
+              delegated_method[:methods].each do |method|
+                next if delegated_method[:to].to_s.match?(/^[@$]/)
+
+                constant_target = if delegated_method[:to] == :class
+                  constant
+                elsif delegated_method[:to].to_s.match?(/^[A-Z]/)
+                  target_klass = constant.const_get(delegated_method[:to])
+                  next unless target_klass.is_a?(Module)
+
+                  target_klass
+                else
+                  false
+                end
+
+                sig = if constant_target
+                  signature_of(constant_target.singleton_method(method))
+                else
+                  signature_of(constant.instance_method(delegated_method[:to]))
+                end
+
+                next unless sig
+
+                delegate_klass = if delegated_method[:allow_nil]
```

Comment:
Yeah - I think this is kind of misnamed - we're just looking up the sig to see if the target returns untyped for method we're delegating so we can do an early return - and after that we don't use it again.